File: web/js/redact.js
// Canvas rendering: the redacted output and the numbered review overlay.

export const STYLES = ["black box", "pixelate", "blur"];

export const CATEGORY_COLORS = {
  secrets: "#dc2626", contact: "#2563eb", financial: "#16a34a", government_id: "#9333ea", network: "#0891b2",
  person: "#ea580c", location: "#ca8a04", dates: "#db2777", faces: "#4f46e5", codes: "#475569", custom: "#059669",
};

const rect = (d) => {
  const x = Math.floor(d.box.x0), y = Math.floor(d.box.y0);
  return [x, y, Math.ceil(d.box.x1) - x, Math.ceil(d.box.y1) - y];
};

/** Draw `source` with every selected detection hidden. */
export function renderRedacted(canvas, source, detections, selected, style = "black box") {
  canvas.width = source.width;
  canvas.height = source.height;
  const ctx = canvas.getContext("2d");
  ctx.drawImage(source, 0, 0);
  const chosen = detections.filter((d) => selected.has(d.id));
  if (!chosen.length) return canvas;

  if (style === "black box") {
    ctx.fillStyle = "#000";
    for (const d of chosen) ctx.fillRect(...rect(d));
    return canvas;
  }

  const minSide = Math.min(canvas.width, canvas.height);
  if (style === "blur" && "filter" in ctx) {
    const radius = Math.max(12, minSide / 40);
    for (const d of chosen) {
      const [x, y, w, h] = rect(d);
      ctx.save();
      ctx.beginPath();
      ctx.rect(x, y, w, h);
      ctx.clip();
      ctx.filter = `blur(${radius}px)`;
      // Blur twice so no glyph shapes survive, then fill the clipped area.
      ctx.drawImage(source, 0, 0);
      ctx.drawImage(canvas, 0, 0);
      ctx.restore();
    }
    return canvas;
  }

  // Pixelate (also the fallback for browsers without canvas filters).
  const block = Math.max(8, Math.round(minSide / 60));
  const tmp = document.createElement("canvas");
  const tctx = tmp.getContext("2d");
  ctx.imageSmoothingEnabled = false;
  for (const d of chosen) {
    const [x, y, w, h] = rect(d);
    tmp.width = Math.max(1, Math.ceil(w / block));
    tmp.height = Math.max(1, Math.ceil(h / block));
    tctx.imageSmoothingEnabled = true;
    tctx.drawImage(source, x, y, w, h, 0, 0, tmp.width, tmp.height);
    ctx.drawImage(tmp, 0, 0, tmp.width, tmp.height, x, y, w, h);
  }
  ctx.imageSmoothingEnabled = true;
  return canvas;
}

/** Draw numbered, colour-coded boxes over `source` for review. */
export function renderReview(canvas, source, detections, selected, draft = null) {
  canvas.width = source.width;
  canvas.height = source.height;
  const ctx = canvas.getContext("2d");
  ctx.drawImage(source, 0, 0);
  const lw = Math.max(2, Math.round(Math.min(canvas.width, canvas.height) / 400));
  const font = Math.max(11, Math.round(Math.min(canvas.width, canvas.height) / 55));
  ctx.font = `600 ${font}px system-ui, sans-serif`;
  ctx.textBaseline = "top";
  for (const d of detections) {
    const on = selected.has(d.id);
    const color = CATEGORY_COLORS[d.category] || "#000";
    const [x, y, w, h] = rect(d);
    if (on) {
      ctx.globalAlpha = 0.28;
      ctx.fillStyle = color;
      ctx.fillRect(x, y, w, h);
      ctx.globalAlpha = 1;
    }
    ctx.setLineDash(on ? [] : [lw * 3, lw * 2]);
    ctx.lineWidth = lw;
    ctx.strokeStyle = color;
    ctx.strokeRect(x, y, w, h);
    ctx.setLineDash([]);
    const label = String(d.id);
    const tw = ctx.measureText(label).width + 8;
    const ty = Math.max(0, y - font - 4);
    ctx.fillStyle = color;
    ctx.fillRect(x, ty, tw, font + 4);
    ctx.fillStyle = "#fff";
    ctx.fillText(label, x + 4, ty + 2);
  }
  if (draft) {
    ctx.setLineDash([6, 4]);
    ctx.lineWidth = lw;
    ctx.strokeStyle = CATEGORY_COLORS.custom;
    ctx.strokeRect(draft.x0, draft.y0, draft.x1 - draft.x0, draft.y1 - draft.y0);
    ctx.setLineDash([]);
  }
  return canvas;
}

File: redactor/redact.py
"""Pixel operations: draw redactions and review annotations."""

from __future__ import annotations

from typing import Iterable

import cv2
import numpy as np

from .types import Detection, Polygon

STYLES = ("black box", "pixelate", "blur")

CATEGORY_COLORS: dict[str, tuple[int, int, int]] = {
    "secrets": (220, 38, 38),
    "contact": (37, 99, 235),
    "financial": (22, 163, 74),
    "government_id": (147, 51, 234),
    "network": (8, 145, 178),
    "person": (234, 88, 12),
    "location": (202, 138, 4),
    "dates": (219, 39, 119),
    "faces": (79, 70, 229),
    "codes": (71, 85, 105),
    "custom": (5, 150, 105),
}


def _mask(shape: tuple[int, ...], polygons: Iterable[Polygon]) -> np.ndarray:
    mask = np.zeros(shape[:2], np.uint8)
    for poly in polygons:
        pts = np.round(np.asarray(poly, np.float32)).astype(np.int32)
        cv2.fillPoly(mask, [pts], 255)
    return mask


def apply_redactions(image: np.ndarray, polygons: list[Polygon], style: str = "black box") -> np.ndarray:
    """Return a copy of ``image`` (H×W×3 RGB uint8) with every polygon hidden."""
    if style not in STYLES:
        raise ValueError(f"style must be one of {STYLES}")
    out = image.copy()
    if not polygons:
        return out
    mask = _mask(out.shape, polygons)
    if style == "black box":
        out[mask > 0] = 0
        return out

    h, w = out.shape[:2]
    if style == "pixelate":
        block = max(8, min(h, w) // 60)
        small = cv2.resize(out, (max(1, w // block), max(1, h // block)), interpolation=cv2.INTER_AREA)
        effect = cv2.resize(small, (w, h), interpolation=cv2.INTER_NEAREST)
    else:
        k = max(31, (min(h, w) // 20) | 1)
        effect = cv2.GaussianBlur(out, (k, k), 0)
        effect = cv2.GaussianBlur(effect, (k, k), 0)
    out[mask > 0] = effect[mask > 0]
    return out


def annotate(image: np.ndarray, detections: list[Detection], selected: set[int] | None = None) -> np.ndarray:
    """Draw numbered, colour-coded outlines; selected detections are also tinted."""
    overlay = image.copy()
    for det in detections:
        if selected is None or det.id in selected:
            pts = np.round(np.asarray(det.polygon, np.float32)).astype(np.int32)
            cv2.fillPoly(overlay, [pts], CATEGORY_COLORS.get(det.category, (0, 0, 0)))
    out = cv2.addWeighted(overlay, 0.3, image, 0.7, 0)
    scale = max(0.4, min(out.shape[:2]) / 1400)
    for det in detections:
        on = selected is None or det.id in selected
        color = CATEGORY_COLORS.get(det.category, (0, 0, 0))
        pts = np.round(np.asarray(det.polygon, np.float32)).astype(np.int32)
        cv2.polylines(out, [pts], True, color, 2 if on else 1, cv2.LINE_AA)
        x, y = int(pts[:, 0].min()), int(pts[:, 1].min())
        label = str(det.id)
        (tw, th), _ = cv2.getTextSize(label, cv2.FONT_HERSHEY_SIMPLEX, scale, 1)
        y_text = max(th + 4, y)
        cv2.rectangle(out, (x, y_text - th - 4), (x + tw + 4, y_text), color, -1)
        cv2.putText(out, label, (x + 2, y_text - 2), cv2.FONT_HERSHEY_SIMPLEX, scale,
                    (255, 255, 255), 1, cv2.LINE_AA)
    return out
